README.md
# castback
Comme laeticia


changelog.txt
====================================================================
Castback addon by [COLOR yellow]@rootsradics and @moacmoa[/COLOR]
====================================================================

v0.0.6 [18-Nov-2018]
- Suppression d'import de modules inutiles
- Bugfix dans la gestion d'une date

v0.0.4 [18-Oct-2017]
- Modification de la référence temporelle pour download json

v0.0.3 [14-Oct-2017]
- Isolating livetv.sx (no more service yet)
- Settings from scratch :
	- Filter streamers
	- JSON url
	- Delay before re-download JSON
- Display only sports and matches with desired links

v0.0.2 [13-Oct-2017]
- Stable version for Live_Sport -> livetv.sx

v0.0.1
-initial release, forked from Castaway


default.py
from __future__ import unicode_literals
from resources.lib.modules.addon import Addon
import sys,os
import urlparse,urllib
import xbmc, xbmcgui, xbmcplugin, xbmcaddon
from resources.lib.modules import control
from resources.lib.modules.log_utils import log
import re

addon = Addon('plugin.video.castback', sys.argv)
addon_handle = int(sys.argv[1])

if not os.path.exists(control.dataPath):
	os.mkdir(control.dataPath)

AddonPath = addon.get_path()
IconPath = os.path.join(AddonPath , "resources/media/")
fanart = os.path.join(AddonPath + "/fanart.jpg")
def icon_path(filename):
	if 'http://' in filename:
		return filename
	return os.path.join(IconPath, filename)

args = urlparse.parse_qs(sys.argv[2][1:])
mode = args.get('mode', None)


# if mode is None:
	
	# sources = os.listdir(AddonPath + '/resources/lib/sources/live_sport')
	# sources.remove('__init__.py')
	# for source in sources:
		# if '.pyo' not in source and '__init__' not in source:
			# try:
				# source = source.replace('.py','')
				# exec "from resources.lib.sources.live_sport import %s"%source
				# info = eval(source+".info()")
				# addon.add_item({'mode': 'open_site_content', 'site': info.mode}, {'title': info.name}, img=icon_path(info.icon), fanart=fanart,is_folder=True)
			# except:
				# pass
	
	# addon.end_of_directory()
	# from resources.lib.modules import cache, control, changelog
	# cache.get(changelog.get, 600000000, control.addonInfo('version'), table='changelog')

	
# elif mode[0] == 'open_site_content':
if mode is None:
	
	# site = args['site'][0]
	site = "livetv"
	
	exec "from resources.lib.sources.live_sport import %s"%site
	info = eval(site+".info()")
	
	source = eval(site+".main()")
	categories  = source.categories()
	for cat in categories:
		addon.add_item({'mode': 'open_sport_cat', 'url': cat[0], 'site': info.mode, 'title': cat[1]}, {'title': cat[1]}, img=icon_path(cat[2]), fanart=fanart,is_folder=True)

	addon.end_of_directory()

elif mode[0]=='open_sport_cat':
	url = args['url'][0]
	site = args['site'][0]
	exec "from resources.lib.sources.live_sport import %s"%site
	info = eval(site+".info()")
	source = eval(site+".main()")
	events = source.events(url)
	for event in events:
		addon.add_item({'mode': 'get_sport_event', 'url': event[0],'site':site , 'title':event[1], 'img': icon_path(info.icon)}, {'title': event[1]}, img=icon_path(info.icon), fanart=fanart,is_folder=True)
		
	
	addon.end_of_directory()	
	
elif mode[0]=='get_sport_event':
	
	url = args['url'][0]
	title = args['title'][0]
	site = args['site'][0]
	img = args['img'][0]
	exec "from resources.lib.sources.live_sport import %s"%site
	info = eval(site+".info()")
	source = eval(site+".main()")
	events = source.links(url)

	autoplay = addon.get_setting('autoplay')
	if autoplay == 'false':
		for event in events:
			browser = 'plugin://plugin.program.chrome.launcher/?url=%s&mode=showSite&stopPlayback=no'%(event[0])
			context = [('Open in browser','RunPlugin(%s)'%browser)]
			addon.add_video_item({'mode': 'play_special_sport', 'url': event[0],'title':title, 'img': img,'site':site}, {'title': event[1]}, img=img, fanart=fanart, contextmenu_items=context)
		addon.end_of_directory()
	else:
		for event in events:
			import liveresolver
			try:
				resolved = liveresolver.resolve(event[0])
			except:
				resolved = None
			if resolved:
				player=xbmc.Player()
				li = xbmcgui.ListItem(title)
				li.setThumbnailImage(img)
				player.play(resolved,listitem=li)
				break
		control.infoDialog("No stream found")

elif mode[0] == 'play_special_sport':
	try:
		url = args['url'][0]
		title = args['title'][0]
		img = args['img'][0]
		site = args['site'][0]
		exec "from resources.lib.sources.live_sport import %s"%(site)
		source = eval(site+'.main()')
		resolved = source.resolve(url)
		li = xbmcgui.ListItem(title, path=resolved)
		li.setThumbnailImage(img)
		li.setLabel(title)
		li.setProperty('IsPlayable', 'true')
		xbmcplugin.setResolvedUrl(int(sys.argv[1]), True, li)
	except:
		log("[ERROR] Erreur dans en mode 'play_special_sport'")
		pass



resources/lib/modules/control.py
# -*- coding: utf-8 -*-

'''
    Genesis Add-on
    Copyright (C) 2015 lambda

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''


import os,xbmc,xbmcaddon,xbmcplugin,xbmcgui,xbmcvfs


lang = xbmcaddon.Addon().getLocalizedString

setting = xbmcaddon.Addon().getSetting

set_setting = xbmcaddon.Addon().setSetting

addon = xbmcaddon.Addon

addItem = xbmcplugin.addDirectoryItem

item = xbmcgui.ListItem

directory = xbmcplugin.endOfDirectory

content = xbmcplugin.setContent

property = xbmcplugin.setProperty

addonInfo = xbmcaddon.Addon().getAddonInfo

infoLabel = xbmc.getInfoLabel

condVisibility = xbmc.getCondVisibility

jsonrpc = xbmc.executeJSONRPC

window = xbmcgui.Window(10000)

dialog = xbmcgui.Dialog()

progressDialog = xbmcgui.DialogProgress()

windowDialog = xbmcgui.WindowDialog()

button = xbmcgui.ControlButton

image = xbmcgui.ControlImage

keyboard = xbmc.Keyboard

sleep = xbmc.sleep

execute = xbmc.executebuiltin

skin = xbmc.getSkinDir()

player = xbmc.Player()

playlist = xbmc.PlayList(xbmc.PLAYLIST_VIDEO)

resolve = xbmcplugin.setResolvedUrl

openFile = xbmcvfs.File

makeFile = xbmcvfs.mkdir

deleteFile = xbmcvfs.delete

listDir = xbmcvfs.listdir

transPath = xbmc.translatePath

skinPath = xbmc.translatePath('special://skin/')

addonPath = xbmc.translatePath(addonInfo('path'))

dataPath = xbmc.translatePath(addonInfo('profile')).decode('utf-8')

settingsFile = os.path.join(dataPath, 'settings.xml')

databaseFile = os.path.join(dataPath, 'settings.db')

favouritesFile = os.path.join(dataPath, 'favourites.db')

sourcescacheFile = os.path.join(dataPath, 'sources.db')

cachemetaFile = os.path.join(dataPath, 'metacache.db')

libcacheFile = os.path.join(dataPath, 'library.db')

metacacheFile = os.path.join(dataPath, 'meta.db')

cacheFile = os.path.join(dataPath, 'cache.db')

jsonFile = os.path.join(dataPath, 'livetvsx.json')


def addonIcon():
    appearance = setting('appearance').lower()
    if appearance in ['-', '']: return addonInfo('icon')
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'icon.png')


def addonPoster():
    appearance = setting('appearance').lower()
    if appearance in ['-', '']: return 'DefaultVideo.png'
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'poster.png')


def addonBanner():
    appearance = setting('appearance').lower()
    if appearance in ['-', '']: return 'DefaultVideo.png'
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'banner.png')


def addonThumb():
    appearance = setting('appearance').lower()
    if appearance == '-': return 'DefaultFolder.png'
    elif appearance == '': return addonInfo('icon')
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'icon.png')


def addonFanart():
    appearance = setting('appearance').lower()
    if appearance == '-': return None
    elif appearance == '': return addonInfo('fanart')
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'fanart.jpg')


def addonNext():
    appearance = setting('appearance').lower()
    if appearance in ['-', '']: return 'DefaultFolderBack.png'
    else: return os.path.join(addonPath, 'resources', 'media', appearance, 'next.jpg')


def artPath():
    appearance = setting('appearance').lower()
    if appearance in ['-', '']: return None
    else: return os.path.join(addonPath, 'resources', 'media', appearance)


def infoDialog(message, heading=addonInfo('name'), icon=addonIcon(), time=3000):
    try: dialog.notification(heading, message, icon, time, sound=False)
    except: execute("Notification(%s,%s, %s, %s)" % (heading, message, time, icon))


def yesnoDialog(line1, line2, line3, heading=addonInfo('name'), nolabel='', yeslabel=''):
    return dialog.yesno(heading, line1, line2, line3, nolabel, yeslabel)


def selectDialog(list, heading=addonInfo('name')):
    return dialog.select(heading, list)


def version():
    num = ''
    try: version = addon('xbmc.addon').getAddonInfo('version')
    except: version = '999'
    for i in version:
        if i.isdigit(): num += i
        else: break
    return int(num)


def refresh():
    return execute('Container.Refresh')


def idle():
    return execute('Dialog.Close(busydialog)')


def queueItem():
    return execute('Action(Queue)')


def openPlaylist():
    return execute('ActivateWindow(VideoPlaylist)')


def openSettings(query=None, id=addonInfo('id')):
    try:
        idle()
        execute('Addon.OpenSettings(%s)' % id)
        if query == None: raise Exception()
        c, f = query.split('.')
        execute('SetFocus(%i)' % (int(c) + 100))
        execute('SetFocus(%i)' % (int(f) + 200))
    except:
        return


def get_keyboard(heading, default=''):
    keyboard = xbmc.Keyboard()
    keyboard.setHeading(heading)
    if default: keyboard.setDefault(default)
    keyboard.doModal()
    if keyboard.isConfirmed():
        return keyboard.getText()
    else:
        return None

IconPath = addonPath + "/resources/media/"
def icon_path(filename):
    return os.path.join(IconPath, filename)

resources/lib/sources/live_sport/streamsports.py
from resources.lib.modules import client,webutils,control
import re,sys,xbmcgui

class info():
    def __init__(self):
        self.mode = 'streamsports'
        self.name = 'Streamsports.me'
        self.icon = 'streamsports.png'
        self.categorized = False
        self.paginated = False
        self.multilink = True

class main():
    def __init__(self):
        self.base = 'http://www.streamsports.io'


    def links(self,url):
        html = client.request(url)
        soup = webutils.bs(html)
        links = soup.findAll('tr')
        links.pop(0)
        links.pop(0)
        links = self.__prepare_links(links)
        return links




    def events(self):
        html = client.request(self.base)
        soup = webutils.bs(html)
        events = soup.findAll('tr')
        events = self.__prepare_events(events)
        return events

    @staticmethod
    def convert_time(time):
        time = time.replace('<hr>','').replace('<br>','')
        li = time.split(':')
        hour,minute=li[0],li[1]
        
        import datetime
        from resources.lib.modules import pytzimp
        d = pytzimp.timezone(str(pytzimp.timezone('Europe/London'))).localize(datetime.datetime(2000 , 1, 1, hour=int(hour), minute=int(minute)))
        timezona= control.setting('timezone_new')
        my_location=pytzimp.timezone(pytzimp.all_timezones[int(timezona)])
        convertido=d.astimezone(my_location)
        fmt = "%H:%M"
        time=convertido.strftime(fmt)
        return time

    def __prepare_events(self,events):
        new = []
        for ev in events:
            try:
                event = ev.findAll('td')
                time = event[0].findAll('span')[1].getText().strip()
                sport = event[1].getText().strip() + '-' +event[2].findAll('span')[1].getText().strip()
                title = event[3].findAll('span')[0].getText() + ' - ' + event[3].findAll('span')[1].getText()
                time = self.convert_time(time)
                url = self.base + event[4].find('a')['href']
                title = '[COLOR orange](%s)[/COLOR] (%s) [B]%s[/B]'%(time,sport,title)
                title = title.encode('utf-8')
                new.append((url,title))
            except:
                pass

        return new

    def __prepare_links(self,links):
        new = []
        precheck = control.setting('link_precheck')
        if precheck=='true':
            pDialog = xbmcgui.DialogProgress()
            pDialog.create('Checking links', 'Checking links...')
        i=1
        items = len(links)
        for l in links:
            try:
                info = l.findAll('td')
                kbps = info[1].getText().strip()
                service = info[2].getText().strip()
                lang = info[3].getText().strip()
                type = info[4].getText().strip()
                url = info[5].find('a')['href']
                title = "%s (%s, %s) - %s"%(service, lang, kbps, type)
                found = True
                perc = 100*i/items
                if type =='HTTP' and precheck=='true':
                    import liveresolver
                    found = liveresolver.find_link(url)
                    pDialog.update(perc, 'Checking:',service)
                if 'sponsored' not in service.lower() and found:
                    new.append((url,title))
                i+=1
                if (precheck=='true' and pDialog.iscanceled()): return new
            except:
                pass
        
        return new
    def resolve(self,url):
        import liveresolver
        return liveresolver.resolve(url,cache_timeout=0)

resources/lib/sources/live_sport/livetv.py
from resources.lib.modules import control
import re,sys,os
from resources.lib.modules.log_utils import log
import json
import urllib
from datetime import datetime, timedelta
import time

class info():
	def __init__(self):
		self.mode = 'livetv'
		self.name = 'Livetv.sx'
		self.icon = 'livetv.png'
		self.categorized = True
		self.paginated = False
		self.multilink = True
		
		
class main():
	def __init__(self):
		self.streamerFilter = []
		self.getStreamerFilter()
		self.json=""
		self.getJson()
	
	def getJson(self):
		if os.path.isfile(control.jsonFile):
			with open(control.jsonFile, "r") as f:
				self.json=f.read()
		if self.json=="" :
			self.downloadJson()
			with open(control.jsonFile, "w") as f:
				f.write(self.json)
		
		self.data = json.loads(self.json)
		
		#Commande "normale" mais ca bug sur mon autre RPI
		#dt=datetime(*(time.strptime(self.data["datetime"], "%Y-%m-%d %H:%M:%S.%f")))
		
		# Du coup, je fais ca :
		j, h=self.data["datetime"].split()
		y, mo, d=j.split("-")
		h, m, s=h.split(":")
		s=s.split(".")[0]
		dt=datetime(int(y), int(mo), int(d), int(h), int(m), int(s))
		
		if datetime.now()>dt+timedelta(minutes=int(control.setting("delay"))):
			self.downloadJson()
			with open(control.jsonFile, "w") as f:
				f.write(self.json)
		
		self.data = json.loads(self.json)
			
		
		
	def getStreamerFilter(self):
		lst=["streamer_acestream", "streamer_sopcast", "streamer_youtube", "streamer_aliez", "streamer_web"]
		for s in lst:
			if control.setting(s)=="true":
				name=s.split("_")[-1]
				self.streamerFilter.append(name)
		
		
	
	def downloadJson(self):
		url = control.setting('livetvsx_json_url')
		response = urllib.urlopen(url)
		self.json=response.read()

	def categories(self):
		cats = self.__prepare_cats()
		return cats
		
	def events(self,sportUrl):
		events = self.__prepare_events(sportUrl)
		return events		

	def links(self,matchUrl):  
		links = self.__prepare_links(matchUrl)
		return links

	def __prepare_cats(self):
		new = []
		for sport in self.data["sports"]:
			matches=self.getDesiredMatches(sport["url"])
			nb=len(matches)
			if nb>0:
				url=sport["url"]
				title="{} ({})".format(sport["name"].encode("utf-8"), nb)
				img=self.getIcon(url)
				new.append((url,title,img))

		return new


	def __prepare_events(self,sportUrl):
		new=[]
		matches=self.getDesiredMatches(sportUrl)
		for match in matches:
			nb=len(self.getDesiredLinks(match["url"]))
			if nb>0:
				strdt=match["datetime"][-8:-3]
				url=match["url"]
				pattern="[COLOR orange][{}][/COLOR] [B]{}[/B] [COLOR green][I]{}[/I][/COLOR] [COLOR blue]({})[/COLOR]"
				title=pattern.format(strdt, match["name"].encode('utf-8'), match["competition"], nb)
				new.append((url,title))
		return new
		
	def __prepare_links(self,matchUrl):
		new = []
		
		links=self.getDesiredLinks(matchUrl)
		for link in links:
			url=link["url"]
			pattern="[COLOR orange][{}][/COLOR] [COLOR blue]({}%)[/COLOR] [B]{}[/B] [COLOR green][I]{}[/I][/COLOR]"
			sInfos="/".join([link["lang"].encode('utf-8'), link["bitrate"].encode('utf-8')]).strip()
			if sInfos!="":
				sInfos=" ["+sInfos+"]"
			title=pattern.format(link["streamer"].encode('utf-8'), link["health"].encode('utf-8'), link["matchname"].encode('utf-8'), link["lang"].encode('utf-8'), sInfos)
			new.append((url,title))
			
		return new
	
	def getDesiredLinks(self, matchUrl):
		for sport in self.data["sports"]:
			res=[m for m in sport["matches"] if m["url"]==matchUrl]
			if len(res)>0:
				match=res[0]
				break
		links=[l for l in match["links"] if l["url"].strip()!="" and l["streamer"] in self.streamerFilter]
		return(links)
		
	def getDesiredMatches(self, sportUrl):
		ret=[]
		sport=[s for s in self.data["sports"] if s["url"]==sportUrl][0]
		for match in sport["matches"]:
			nb=len(self.getDesiredLinks(match["url"]))
			if nb>0:
				ret.append(match)
		return(ret)
	
	def resolve(self,url):
		import liveresolver
		return liveresolver.resolve(url,cache_timeout=0)
		
	def getIcon(self, sportUrl):
		table={
			"1":	"soccer.png",
			"2":	"hockey.png",
			"3":	"basketball.png",
			"4":	"tennis.png",
			"5":	"volleyball.png",
			"6":	"fighting.png",
			"7":	"f1.png",
			"9":	"athletics.png",
			"12":	"soccer.png",
			"13":	"handball.png",
			"17":	"rugby.png",
			"18":	"skiing.png",
			"19":	"baseball.png",
			"20":	"curling.png",
			"21":	"volleyball.png",
			"27":	"football.png",
			"29":	"snooker.png",
			"30":	"darts.png",
			# "32":	"floorball.png",
			"33":	"rugby.png",
			"37":	"golf.png",
			"38":	"hockey.png",
			"39":	"table_tennis.png",
			"40":	"cycling.png",
			"41":	"cricket.png",
			"54":	"tennis.png",
			"59":	"hockey.png",
			"72":	"horse_racing.png",
			# "74":	"esport.png",
			"75":	"fighting.png",
			# "93":	"climbing.png",
			"96":	"volleyball.png",
		}
		id=sportUrl.split("/")[-2]
		if id in table:
			return("icons/"+table[id])
		else:
			return("live_sport.jpg")